jmxtrans-output/jmxtrans-output-influxdb/src/main/java/com/googlecode/jmxtrans/model/output/InfluxDbWriter.java

```diff
@@ -0,0 +1,307 @@
+
+	/**
+	 * The deault <a href=
+	 * "https://influxdb.com/docs/v0.9/concepts/key_concepts.html#retention-policy">
+	 * The retention policy</a> for each measuremen where no retentionPolicy
+	 * setting is provided in the json config
+	 */
+	private static final String DEFAULT_RETENTION_POLICY = "default";
+
+	private String database;
+	private ConsistencyLevel writeConsistency = ConsistencyLevel.ALL;
+	private String retentionPolicy;
+
+	/** Thread safe **/
+	InfluxDB influxDB;
+
+	/**
+	 * @param typeNames
+	 * @param booleanAsNumber
+	 * @param debugEnabled
+	 * @param url
+	 *            - The url e.g http://localhost:8086 to InfluxDB
+	 * @param username
+	 *            - The username for InfluxDB
+	 * @param password
+	 *            - The password for InfluxDB
+	 * @param database
+	 *            - The name of the database (created if does not exist) on
+	 *            InfluxDB to write the measurements to
+	 * @param settings
+	 */
+	@JsonCreator
+	public InfluxDbWriter(@JsonProperty("typeNames") ImmutableList<String> typeNames,
+			@JsonProperty("booleanAsNumber") boolean booleanAsNumber, @JsonProperty("debug") Boolean debugEnabled,
+			@JsonProperty("url") String url, @JsonProperty("username") String username,
+			@JsonProperty("password") String password, @JsonProperty("database") String database,
+			@JsonProperty("settings") Map<String, Object> settings) {
+		super(typeNames, booleanAsNumber, debugEnabled, settings);
+
+		this.database = database;
+
+		initWriteConsistency(settings);
+		initRetentionPolicy(settings);
+		initResultAttributesToWriteAsTags(settings);
+
+		LOG.debug("Connecting to url: {} as: username: {}", url, username);
+
+		influxDB = InfluxDBFactory.connect(url, username, password);
+	}
+
+	private void initWriteConsistency(Map<String, Object> settings) {
+		String consistencySetting = Settings
+				.getStringSetting(settings, SETTING_WRITE_CONSISTENCY, DEFAULT_WRITE_CONSISTENCY).toUpperCase();
+		writeConsistency = ConsistencyLevel.valueOf(consistencySetting);
+		LOG.debug("Write consistency set to: {}", writeConsistency);
+	}
+
+	private void initRetentionPolicy(Map<String, Object> settings) {
+		retentionPolicy = Settings.getStringSetting(settings, SETTING_RETENTION_POLICY, DEFAULT_RETENTION_POLICY);
+		LOG.debug("Retention Policy set to: {}", retentionPolicy);
+	}
+	
+	private void initResultAttributesToWriteAsTags(Map<String, Object> settings) {
+		@SuppressWarnings("unchecked")
+		List<String> resultTagsFromSettings = (List<String>) settings.get(SETTING_RESULT_TAGS);
+		if (resultTagsFromSettings != null) {
+			resultAttributesToWriteAsTags.clear();
+			for (String resultTag : resultTagsFromSettings) {
+				resultAttributesToWriteAsTags.add(ResultAttribute.valueOf(resultTag.toUpperCase()));
+			}
+		}
+		LOG.debug("Result Tags to write set to: {}", resultAttributesToWriteAsTags);
+	}
+
+	@VisibleForTesting
+	void setInfluxDB(InfluxDB influxDB) {
```

Comment:
The fact that GraphiteOutputWriter2 can still be constructed directly does not defeat the purpose of a factory. What I'm trying to do is separate the construction logic from the execution logic. It makes testing and composition easier. In the case of InfluxBDOutputWriter, basically everything that is in its current constructor should be moved to a `InfluxDBOutputWriterFactory` class and the runtime behaviour should be kept in a `InfluxDBOutputWriter` class. I'll try to do a PR on your PR later this evening to explain better...